### ISSUE TITLE
Start/Stop-Transcript does not capture everything written to the console

### ISSUE THREAD

**Issue by zammitt** [PowerShell/PowerShell#10994]
Start/Stop-Transcript does not always capture everything written to the console.

I created a github repo with code, repro steps and more context: https://github.com/zammitt/PowerShellTranscriptIssue

# Steps to reproduce

In a script file (.ps1)
1) Start-Transcript
2) Run a Cmdlet that outputs data with WriteObject
3) Stop-Transcript
4) Observe that the output was not written to the transcript

Here is a script to repro:

```
Install-Module TranscriptIssue
Import-Module TranscriptIssue

Start-Transcript -Path transcript.txt -UseMinimalHeader

Write-Output "Hello from Write-Output"
Write-Host "Hello from Write-Host"

Test-TranscriptIssue

Write-Output "Goodbye from Write-Output"
Write-Host "Goodbye from Write-Host"

Stop-Transcript

# Write out the transcript to the console
cat ./transcript.txt
```
The code for Test-TranscriptIssue can be found here:  https://github.com/zammitt/PowerShellTranscriptIssue/blob/master/TestIssueOnPowerShellCore/Test-TranscriptIssue.cs

# Expected behavior

The contents of the transcript match what was written to the console while the transcript was active.

```
Transcript started, output file is transcript.txt
Hello from Write-Output
Hello from Write-Host

Goodbye from Write-Host
TestProperty
------------
TestValue
Goodbye from Write-Output
Transcript stopped, output file is D:\code\PowerShellTranscriptIssue\TestIssueOnPowerShellCore\bin\Debug\netcoreapp3.0\transcript.txt
**********************
PowerShell transcript start
Start time: 20191105180138
**********************
Hello from Write-Output
Hello from Write-Host

Goodbye from Write-Host
TestProperty
------------
TestValue
Goodbye from Write-Output
**********************
PowerShell transcript end
End time: 20191105180138
**********************

```

# Actual behavior

The transcript is missing data that was written to the console while the transcript was active.

```
Transcript started, output file is transcript.txt
Hello from Write-Output
Hello from Write-Host

Goodbye from Write-Host
TestProperty
------------
TestValue
Goodbye from Write-Output
Transcript stopped, output file is D:\code\PowerShellTranscriptIssue\TestIssueOnPowerShellCore\bin\Debug\netcoreapp3.0\transcript.txt
**********************
PowerShell transcript start
Start time: 20191105180138
**********************
Transcript started, output file is transcript.txt
Hello from Write-Output
Hello from Write-Host

Goodbye from Write-Host
**********************
PowerShell transcript end
End time: 20191105180138
**********************

```

# Environment data

```
Name                           Value
----                           -----
PSVersion                      6.2.3
PSEdition                      Core
GitCommitId                    6.2.3
OS                             Microsoft Windows 10.0.17134
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by JamesWTruher** [comment]
TLDR:
* Mixing and matching objects, strings, and write-host with transcription is _really_ tricky.
 
I’m not sure this is a bug, per se, but a subtle interaction between the script and how transcription works with our pipelines. I will agree that it is in no way intuitive.
 
One of the things that happen when you emit objects is that the formatting of those objects gets handled by out-default, but after the script has completed and since transcription has stopped, that output doesn’t get transcribed. Strings are handled differently, they’re sometimes short-circuited through formatting (but sometimes not). The combination of write-host (which makes an immediate write), write-object <string> (which may get short-circuited through formatting), and output of a <complexObject> makes this pretty hard to reason about what’s going to happen. However, there may be a way that you may want to try (See “A POSSIBLE WAY” below).
 
Even more details:
 
Here are a couple of variations on the theme:
 
Variation 1 – your scenario, with the output of a structured object at the end of all of the other operations.

```powershell 
PS> cat s2.ps1
remove-item s2 -ea ignore
start-transcript s2 -usemin
write-host "1"
write-output "2"
get-location
write-host "4"
write-output "5"
stop-transcript
 
PS> ./s2.ps1
Transcript started, output file is s2
1
2
 
4
Path
----
/Users/jimtru/src/projects/transcript
5
Transcript stopped, output file is /Users/jimtru/src/projects/transcript/s2
 ```
 
This one is really confusing, but explainable (but perhaps not satisfying). ‘Write-Host 4’ is visible before `get-location` because Write-Host is hot-rodded directly into the host and emitted (there’s code in transcription which essentially tees the output to the transcript file and the console). Then we have the regular output of `get-location` and `write-output 5` sent as output of the script.

Since transcription is turned off before the script exits, it’s not rendered in the transcript, because the objects were sent to the next consumer in the pipeline (in this case, it’s out-default which is what the engine does automatically). The tricky bit is that strings are also somewhat hot-rodded in our formatting, so the first `write-output 2` gets emitted and captured by the transcript, but the insertion of the `get-location` object causes it to be pushed into the stack of things that need actual formatting, which sets a bit of state for any remaining objects which may need formatting (which is why the second `write-output 5` doesn’t get transcripted.
 
 ```powershell
**********************
PowerShell transcript start
Start time: 20191106114858
**********************
Transcript started, output file is s2
1
2
 
4
**********************
PowerShell transcript end
End time: 20191106114858
**********************
 ```

Variation 2 – move the object emission to the beginning

```powershell 
start-transcript s3 -usemin
get-location
write-host "1"
write-output "2"
write-host "4"
write-output "5"
stop-transcript
```

We can see that the `write-host` happen before anything, and then the objects start to come out. The string output essentially forces the object to be rendered to the screen, but you’ll notice that the transcript contains only the output of Write-Host. That’s because those were piped to the formatting after the script turned off transcription.
 
```powershell 
PS> ./s3.ps1
Transcript started, output file is s3
 
1
4
Path
----
/Users/jimtru/src/projects/transcript
2
5
Transcript stopped, output file is /Users/jimtru/src/projects/transcript/s3
 
  
PS> cat s3
 
**********************
PowerShell transcript start
Start time: 20191106115142
**********************
Transcript started, output file is s3
 
1
4
**********************
PowerShell transcript end
End time: 20191106115142
**********************
```

Variation 3 – object emission at the end:
We can see that everything but the object is now in the transcript
 
```powershell
PS> cat s4.ps1
start-transcript s4 -usemin
write-host "1"
write-output "2"
write-host "4"
write-output "5"
get-location
stop-transcript
```
 
Here’s the execution flow

```powershell 
PS> ./s4.ps1
Transcript started, output file is s4
1
2
4
5
 
Path
----
/Users/jimtru/src/projects/transcript
Transcript stopped, output file is /Users/jimtru/src/projects/transcript/s4
```
 
And the content of the transcript

```powershell
**********************
PowerShell transcript start
Start time: 20191106115255
**********************
Transcript started, output file is s4
1
2
4
5
 
**********************
PowerShell transcript end
End time: 20191106115255
**********************
```
 
## A Possible Way

Lastly, here’s a slight variation on the original, but now everything will be in the transcript. I’ve taken the code of interest and am calling the formatter forcefully via out-default. You’ll notice that the last Write-Host call is still out of order, that’s because of the short-circuiting of write-host which is a direct call into the host and does not go into the output stream.

```powershell 
start-transcript s5 -usemin
.{
    write-host "1"
    write-output "2"
    get-location
    write-host "4"
    write-output "5"
} | out-default
stop-transcript
```
 
Here’s the execution, followed by the transcript
 
```powershell
PS> ./s5.ps1
Transcript started, output file is s5
1
2
 
4
Path
----
/Users/jimtru/src/projects/transcript
5
 
Transcript stopped, output file is /Users/jimtru/src/projects/transcript/s5
  
 
PS> cat s5
**********************
PowerShell transcript start
Start time: 20191106115701
**********************
Transcript started, output file is s5
1
2
 
4
Path
----
/Users/jimtru/src/projects/transcript
5
 
**********************
PowerShell transcript end
End time: 20191106115701
**********************
```

**Comment by he852100** [comment]
I suggest adding a new option. Allows it to output color documents

**Comment by zammitt** [comment]
@JamesWTruher Thank you for the detailed reply.  Out-Default works great!   I ended up using the -Transcript flag so that the objects are still passed forward on the pipeline.

@he852100 I'll second this.  Color in transcripts would be awesome.

**Comment by vexx32** [comment]
My one question to that is always going to be -- how? PS works with text files, not formatted documents. The next step up would perhaps be Markdown, which would be messier to read in the console, and still doesn't natively support color formatting. That should probably be a whole other issue discussion, though. 😉

**Comment by vexx32** [comment]
My one question to that is always going to be -- how? PS works with text files, not formatted documents. The next step up would perhaps be Markdown, which would be messier to read in the console, and still doesn't natively support color formatting. That should probably be a whole other issue discussion, though. 😉

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this, we believe a complete solution for transcription (like the unix `script` command) would love to see the community build a new cmdlet (published to PSGallery) that uses ptty or on Windows grabbing the screen buffer, but the ask is out of scope for the current transcription framework.

**Comment by Cloudmersive** [comment]
Dear Microsoft - when are you going to fix this?  @SteveL-MSFT this is critical functionality - how is it "out of scope"?  It is critically important that bedrock Microsoft capabilities actually solve problems at scale and this is a huge gap

**Comment by mklement0** [comment]
To complete the picture of the current problems and limitations:

* Interactive prompts are never transcribed - unlike with the Unix `script` utility, you neither see the prompt string nor what the user typed.

* #4645 describes duplicated output in transcripts and situationally _changed behavior_ when transcript ion is used.

* As for the `Out-Default` workaround for the issue at hand:
  * Given that we [shouldn't promote use of `Out-Default` in user code](https://github.com/MicrosoftDocs/PowerShell-Docs/issues/5131), `Out-Host` is the better choice, but it is still a flawed workaround:

       *  You lose the ability to capture any output from the script.
       * Due to #11134, stderr output from _external programs_ will be missing from the transcript.

On Unix-like platforms, use of the [`script` utility](http://man7.org/linux/man-pages/man1/script.1.html) already is an option to bypass all these issues (albeit not from _within_  a script), while additionally providing:

* capturing of VT sequences (ANSI escapes) 
* capturing of interactive prompt strings and _what the user typed_

However, the `PSReadLine` module doesn't play nice with the latter in `Read-Host` calls; e.g. (on Linux):

```sh
# Transcribe an interactive PowerShell user prompt.
# Save the transcript in file .\typescript
script -c 'pwsh -nop -c "read-host prompt"'
```

If you type `abc` when prompted, the transcript won't contain just string `abc`, but equivalent following, which makes the terminal go haywire when you `cat` the resulting file; note how each letter typed caused the inclusion of VT sequences and reprinting of all letters typed so far:

```
`e[?1h`e=prompt:·`e[6na`e[62;9Hab`e[62;9Habc
```

**Comment by sba923** [comment]
Just discovered that `Start-Transcript` typically captures little to no output from `ROBOCOPY`.

![image](https://user-images.githubusercontent.com/12860484/101187692-f9247180-3654-11eb-903e-48347af5a4e3.png)

![image](https://user-images.githubusercontent.com/12860484/101187961-5ae4db80-3655-11eb-83ba-188f19f79f53.png)

![image](https://user-images.githubusercontent.com/12860484/101188010-6932f780-3655-11eb-9298-c495e33a088d.png)


[PSTranscriptAndRobocopy.zip](https://github.com/PowerShell/PowerShell/files/5643944/PSTranscriptAndRobocopy.zip)

Would there be a workaround?

**Comment by mklement0** [comment]
I see the problem in 7.1.0 (as you do), and also still in in 7.2.0-preview.1.

Appending `| Out-Host` is a workaround for PS Core (none needed for WinPS), though the formatting is a bit off.

**Comment by aolszowka** [comment]
This also does not work for `WhatIf` if ran in an Invoke-Command; See https://github.com/PowerShell/PowerShell/issues/9870#issuecomment-856969529

**Comment by kilasuit** [comment]
I was playing with .Net Interactive notebooks & looking into this for why this issue with Notebooks occurs https://github.com/dotnet/interactive/issues/1430 as this is a crucial item that needs fixed before I can recommend to Jupyter Notebooks using .Net Interactive to users in production environments that will have org wide transcription enforced, & I would like to be able to recommend using them without needing to resort to a non-workable workaround cc @SteveL-MSFT

**Comment by Zeal0us** [comment]
This is pretty disappointing, it's one of those things you expect to just work when you find it, and then you check your log file and it's got 20 lines out of tens of thousands.

**Comment by TheOnlyWei** [comment]
Same problem when you try to run PowerShell 7 code with Start-Transcript from a PowerShell 5 session. You see the output in PowerShell 5, but when you check the log file that Start-Transcript output to, it is missing a lot of the output.

**Comment by Cistalvo** [comment]
Any dev update on this? I might have stumbled over this error, just wondering if it was fixed in the meantime.

**Comment by buscseik** [comment]
This issue still exists and it is worse than ever. It was not good in PS5 but in PS7 even worse. 
Start-transcript can't literally be used 

This is a crutial feature. Please fix it. This issue started before 2015.
PowerShell is great, Windows Terminal is great, but due this I will have to switch some third party solution like Moba Xterm which actually can log everthing without any issue.

**Comment by johnbailey-transactcampus** [comment]
@SteveL-MSFT - This issue is still prevalent. I can reproduce the `Start-Transcript` issue if the `Invoke-RestMethod` commandlet is called with a `POST` method. 

In my case, the workaround of utilising `| Out-Host` (mentioned above) on PS 7.2.5 still does not produce any viable command invocation output (for monitoring/auditing) - whether the command succeeds **_or_** fails:

```
**********************
Command start time: 20220722090610
**********************
PS>TerminatingError(Invoke-RestMethod): "{ "statusCode": 401, "message": "Unauthorized. Access token is missing or invalid." }"
```

If I append `-Verbose` and `-Debug` on the Invoke-RestMethod command, then I lose the Invocation Header data (e.g.: `Command start time:`) from the transcript for the command - which makes it even harder to determine the details of the command that was run (when a failure doesn't occur):

```
VERBOSE: DELETE with 0-byte payload
VERBOSE: received 85-byte response of content type application/json
>> TerminatingError(Invoke-RestMethod): "{ "statusCode": 401, "message": "Unauthorized. Access token is missing or invalid." }"
```

As it appears that there hasn't been any traction on this - for quite some time - is it possible to help the community derive a work-around for the[se] issue[s], until it[they] can be addressed in the code-base?

**Comment by SteveL-MSFT** [comment]
I hope people understand that if this was easy to fix, we would have done it.  The transcription code is more than 10 years old and its design has limitations not easily addressed.  The robocopy example above is an example where the NativeCommandProcessor has an optimization where if a native command is at the end of the pipeline, then it doesn't read the stdout pipe and instead just has the exe send the output directly to the console.  In this case, transcription is not aware (since PowerShell is not aware) of any output.  Piping to a cmdlet works around this optimization.

It would actually be easier to create a `script`-like executable (perhaps called `transcript.exe` that does what `script` does which is start a new instance of the shell that ran it and read stdout/stderr/stdin and write that to a file while also passing it through to the shell.  Doing it within the shell is problematic due to output being "leaky".

**Comment by psawa** [comment]
Then what's the point of this command if the transcription cannot be relied on? This is just an extreme flaw for a crucial and basic cmdlet

**Comment by Xechorizo** [comment]
> if this was easy to fix, we would have done it

If it weren't vital functionality, why would the community want it? Your team only fixes easy problems?

If it can't be done, remove it.

**Comment by sba923** [comment]
> If it can't be done, remove it.

I'd rather have PowerShell with limited transcript functionality than with none at all.

**Comment by nzbart** [comment]
While I agree that `Start-Transcript` is fairly useless due to this problem and not capturing the output from executables, if anyone has a workaround for this problem, please post it here. A suitable workaround could include something like:

1. A new function such as recommended here: https://github.com/PowerShell/PowerShell/issues/10994#issuecomment-572306095; or
2. A new host process that can "tee" the output to file.

@SteveL-MSFT currently, the documentation for `Start-Transcript` says:

>The transcript includes all command that the user types and all output that appears on the console.

Would it be possible to at least list the limitations of the current command, such as not capturing the output from executables in the documentation to save people a lot of time wondering why it's not working as documented?

**Comment by Bibench-R7** [comment]
It does capture executable output, but only 10 or so on hundreds. Since it looks like it works for a few lines test, I discovered after the fact that all my logs are worthless.

**Comment by SyntevoAlex** [comment]
My understanding of the issue is that this is a race. If PowerShell doesn't "flush" output before `Stop-Transcript` is executed, the output gets lost. In my scripts, I think I'm seeing that sometimes output lands, and sometimes it's lost.

Regarding workarounds: is it possible to force PowerShell to flush? I would just do it before `Stop-Transcript`, problem solved?

**Comment by aetonsi** [comment]
Ok i just discovered the reason why the transcript for `choco list -localonly -all` saves **literally** 9 out of 151 lines of output....

And it saves absolutely **nothing** if i pipe to grep (`choco list -localonly -all | grep -i a`)

**Comment by aetonsi** [comment]
> My understanding of the issue is that this is a race. If PowerShell doesn't "flush" output before `Stop-Transcript` is executed, the output gets lost. In my scripts, I think I'm seeing that sometimes output lands, and sometimes it's lost.
> 
> Regarding workarounds: is it possible to force PowerShell to flush? I would just do it before `Stop-Transcript`, problem solved?

Mmh that's not a bad idea, if that is really the culprit.. so far i tried the following but it didn't do anything, maybe someone knows something more?
```powershell
PS>[console]::OpenStandardOutput().Flush()
PS>[console]::OpenStandardError().Flush()
PS>[console]::Out.Flush()
PS>[console]::Error.Flush()
```

**Comment by sba923** [comment]
Yesterday found out that `Start-Transcript` also does not capture the output from VScode command-line operations such as `code --install-extension foo`. One more case that makes it essentially useless...

**Comment by DennisL68** [comment]
This project claims to redirect all PowerShell output to the console to a transcript file...

https://github.com/dlwyatt/PowerShellLoggingModule

However, it hasn't been updated in a while and the latest release on PS Gallery isn't based on the latest source.

https://www.powershellgallery.com/packages/PowerShellLogging

**Comment by aetonsi** [comment]
is it fixed at least?

**Comment by triumph1** [comment]
not fixed yet...???

**Comment by poa00** [comment]
> This project claims to redirect all PowerShell console output to a transcript file...
> 
> https://github.com/dlwyatt/PowerShellLoggingModule
> 
> However, it hasn't been updated in a while and the latest release on PS Gallery isn't based on the latest source.
> 
> https://www.powershellgallery.com/packages/PowerShellLogging
> 
> > <img alt="" width="48" src="https://camo.githubusercontent.com/8300fcf7d9be98f5cffb35c5a7342f293385dc7d2ac2413f2463aa4aabc7db63/68747470733a2f2f6f70656e67726170682e6769746875626173736574732e636f6d2f386636623366326635336134643637663131656135353364363131306434383862653734343062373832346139333739663636333963653338383935386531612f646c77796174742f506f7765725368656c6c4c6f6767696e674d6f64756c65"><img alt="" height="14" src="https://camo.githubusercontent.com/95e35b1ddbd18e9680ef4876f3544fdf46e2116b5fefc3f6a48e1e427ea75b59/68747470733a2f2f6769746875622e6769746875626173736574732e636f6d2f66617669636f6e732f66617669636f6e2e737667"> GitHub**[GitHub - dlwyatt/PowerShellLoggingModule: Automatically capture and save PowerShell console output from a script (without the limitations and headaches of Start-Transcript)](https://github.com/dlwyatt/PowerShellLoggingModule)**Automatically capture and save PowerShell console output from a script (without the limitations and headaches of Start-Transcript) - GitHub - dlwyatt/PowerShellLoggingModule: Automatically capture ...
> 
> > <img alt="" width="48" src="https://camo.githubusercontent.com/86c9d1d83d76d6ddfd2a83e82737c850ad2d5bc517ba070e0bf52122cdb49a8a/68747470733a2f2f706f7765727368656c6c67616c6c6572792e636f6d2f436f6e74656e742f496d616765732f4272616e64696e672f5061636b61676544656661756c7449636f6e2e706e67">**[PowerShellLogging 1.3.0](https://powershellgallery.com/packages/PowerShellLogging/)**Captures PowerShell console output to a log file.
> 
> The about_PowerShellLogging file warns about this module breaking easily as it is using private members of the PowerShell classes

There do not appear to be many viable (actively maintained) workarounds on GitHub at all from what I can find, but of the few available, the above does appear to be in functioning condition and with the _relatively_ highest level of engagement.  I haven't tried it yet but will report back once I do.  
***
https://github.com/dlwyatt/PowerShellLoggingModule

**Comment by jkmartindale** [comment]
mfw I'm still running `script` in WSL and then running powershell.exe in that session because I can't trust Start-Transcript

**Comment by arakielG** [comment]
Just to add my voice to this still being an issue.  I have an install script that runs a bunch of winget commands.  Start-Transcript captures the first half dozen or so and then after that the only thing getting transcribed is the Write-Hosts I have between them all
`==================================
===STARTING APPLICATION INSTALL===
==================================
Installing EdgeView2Runtime...
Found an existing package already installed. Trying to upgrade the installed package...
No available upgrade found.
No newer package versions are available from the configured sources.
Installing Google Drive...
Found an existing package already installed. Trying to upgrade the installed package...
No available upgrade found.
No newer package versions are available from the configured sources.
Installing Google Chrome...
Found an existing package already installed. Trying to upgrade the installed package...
No available upgrade found.
No newer package versions are available from the configured sources.
Installing BitWarden...
Found an existing package already installed. Trying to upgrade the installed package...
No available upgrade found.
No newer package versions are available from the configured sources.
Installing Git...
Found an existing package already installed. Trying to upgrade the installed package...
No available upgrade found.
No newer package versions are available from the configured sources.
Installing GitLFS...
Found an existing package already installed. Trying to upgrade the installed package...
No available upgrade found.
No newer package versions are available from the configured sources.
Installing GitHub Desktop...
Found an existing package already installed. Trying to upgrade the installed package...
No available upgrade found.
No newer package versions are available from the configured sources.
Installing UnityHub...

Installing JetBrains Rider...

Installing VSCodium...

Installing Windows Terminal...

Installing Powershell...

Installing WSL...

Installing Ubuntu...

` ... etc

**Comment by A9G-Data-Droid** [comment]
The response from Microsoft here has been really surprising. At the very least, `Stop-Transcript` should call for a flush and wait for flush of anything at the point when stop is called. As for it's inability to capture all output, that is certainly a bug and it's absurd to say it's not. There should be no workaround or call upon the community. This feature doesn't do what it says on the tin. Is it supported or not?

**Comment by sba923** [comment]
IMVHO a broken transcript feature is about as useful as no transcript feature.

So, this must be fixed, or removed/deprecated.

**Comment by jkmartindale** [comment]
still a footgun, bot bro

**Comment by DennisL68** [comment]
Just because no one is working actively on this issue doesn't mean that a solution isn't in high demand.

**Comment by lonbluster** [comment]
broken on Windows 2019, no output generated in transcript file
PSVersion                      5.1.17763.8146

**Comment by jkmartindale** [comment]
> This issue has not had any activity in 6 months, if there is no further activity in 7 days, the issue will be closed automatically.
> 
> Activity in this case refers only to comments on the issue. If the issue is closed and you are the author, you can re-open the issue using the button below. Please add more information to be considered during retriage. If you are not the author but the issue is impacting you after it has been closed, please submit a new issue with updated details and a link to this issue and the original.

Bump

**Comment by EclipsedStar0** [comment]
Powershell transcription is completely missing over half the console output of running a python file printing to console... and they are all strings. Which was said to be done correctly. No it ain't. This is broken.
